lib/tapioca/executor.rb

```diff
@@ -38,11 +44,29 @@ def run_in_parallel(&block)
       workers = (0...@number_of_workers).map do
         items = @queue.shift(@items_per_worker)
 
-        fork { block.call(items.shift) until items.empty? }
+        fork do
+          read.close
+          result = items.map { |item| block.call(item) }.compact
+
+          packed = [Marshal.dump(result)].pack("m")
+          write.puts("-#{packed}") unless result.empty?
+        end
       end
 
+      write.close
+
       # Wait until all the workers finish
       workers.each { |pid| Process.waitpid(pid) }
+
+      result = read.read
+      read.close
+
+      if result
+        result
+          .split("-")
+          .drop(1)
+          .flat_map { |item| T.unsafe(Marshal.load(item.unpack1("m"))) }
```

Comment:
It would be great to add comment here as to what this is really doing. The mechanics are easy to understand but the why is a little harder. Extracting the delimiter will help a little with the explanation too.

Maybe something like:
```suggestion
        result
          .split("-") # Split the output at delimiter that separates each worker's output
          .drop(1) # Drop the leading empty space
          .flat_map { |item| T.unsafe(Marshal.load(item.unpack1("m"))) } # Unpack the data and merge
```

Reply:
Added some explanation before these invocations. Let me know what you think.